internal/langserver/handlers/did_change_watched_files.go

```diff
@@ -0,0 +1,181 @@
+package handlers
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/creachadair/jrpc2"
+	"github.com/hashicorp/terraform-ls/internal/document"
+	"github.com/hashicorp/terraform-ls/internal/job"
+	"github.com/hashicorp/terraform-ls/internal/protocol"
+	lsp "github.com/hashicorp/terraform-ls/internal/protocol"
+	"github.com/hashicorp/terraform-ls/internal/terraform/ast"
+	"github.com/hashicorp/terraform-ls/internal/uri"
+)
+
+func (svc *service) DidChangeWatchedFiles(ctx context.Context, params lsp.DidChangeWatchedFilesParams) error {
+	var ids job.IDs
+
+	for _, change := range params.Changes {
+		rawURI := string(change.URI)
+
+		rawPath, err := uri.PathFromURI(rawURI)
+		if err != nil {
+			svc.logger.Printf("error parsing %q: %s", rawURI, err)
+			continue
+		}
+
+		if change.Type == protocol.Deleted {
+			// We don't know whether file or dir is being deleted
+			// 1st we just blindly try to look it up as a directory
+			_, err = svc.modStore.ModuleByPath(rawPath)
+			if err == nil {
+				svc.removeIndexedModule(ctx, rawURI)
+				continue
+			}
+
+			// 2nd we try again assuming it is a file
+			parentDir := filepath.Dir(rawPath)
+			_, err = svc.modStore.ModuleByPath(parentDir)
+			if err != nil {
+				svc.logger.Printf("error finding module (%q deleted): %s", parentDir, err)
+				continue
+			}
+
+			// and check the parent directory still exists
+			fi, err := os.Stat(parentDir)
+			if err != nil {
+				if os.IsNotExist(err) {
+					// if not, we remove the indexed module
+					svc.removeIndexedModule(ctx, rawURI)
+					continue
+				}
+				svc.logger.Printf("error checking existence (%q deleted): %s", parentDir, err)
+				continue
+			}
+			if !fi.IsDir() {
```

Comment:
Curious as to why this check is here since we're doing this earlier:
```
parentDir := filepath.Dir(rawPath)
```

Reply:
Right, this should basically never happen - I was just trying to be a bit more defensive here too.

The only realistic scenarios which come to mind involve cases where it's not a file nor a directory, but e.g. symlink or any special type of file - which is mostly relevant on Unix-based systems:
https://pkg.go.dev/io/fs#FileMode

Reply:
Okay, so just yesterday, I got bit by a severe production bug wherein I was copying files from one directory to another. Without a safety check of whether the file is a regular file or not, we ended up doing a cp on a file which was a UNIX pipe and the application just froze indefinitely waiting for the copy to finish would _never_ be the case. I can totally understand the need for this check here howsoever obvious it might seem.